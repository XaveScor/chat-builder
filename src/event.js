/* @flow */
type Watcher<T> = (arg: T) => void;

export type EventType<T> = {
    (T): void,
    watch(Watcher<T>): () => void,
    waitMessage(): Promise<T>,
    lastMessage(): T | void,
};

export function createEvent<T>(): EventType<T> {
    const watchers: Array<Watcher<T>> = [];

    let lastMessage: T | void
    function Event(arg: T) {
        lastMessage = arg
        watchers.forEach(f => f(arg))
    }
    Event.watch = (subscriber: Watcher<T>) => {
        watchers.push(subscriber);
        const id = watchers.length - 1;
        return () => {
            delete watchers[id];
        }
    };
    Event.waitMessage = async () => {
        return new Promise(resolve => {
            const unsubscribe = Event.watch(data => {
                resolve(data)
                unsubscribe()
            })
        })
    }
    Event.lastMessage = () => lastMessage

    return Event;
}

export type WaitMessage<Event> = $ElementType<Event, 'waitMessage'>
