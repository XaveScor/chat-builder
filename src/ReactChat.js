// @flow

import * as React from 'react'
import {Page} from './createPage'
import type {RunChat} from './runChat'
import {createEvent} from './event'
import type {NotifyViewEvent, State, PendingConfig, BubbleContainer} from './types'
import {context} from './BubbleEnd'

type DialogProps = {
    views: $PropertyType<State, 'dialog'>,
    pending?: $PropertyType<PendingConfig, 'pending'>,
    bubbleContainer?: BubbleContainer,
    bubbleEnd: React.Node,
}
const Dialog = ({views, pending: Pending, bubbleContainer: Container, bubbleEnd}: DialogProps) => {
    const list = Array.from(
        views.map((v, idx) => <v.component {...v.props} key={idx} />).values()
    )
    const totalView = <>
        {list}
        {bubbleEnd}
    </>

    if (Container != null) {
        return <Container>{totalView}</Container>
    }

    return totalView
};

type Props<TProps> = {
    runChat: RunChat,
    page: Page<TProps>,
    pending?: PendingConfig,
    bubbleContainer?: BubbleContainer,
}
export const Chat = <TProps>(props: Props<TProps>) => {
    const {page, pending, runChat, bubbleContainer} = props
    const [data, setData] = React.useState<State | void>();
    const [stopped, setStopped] = React.useState<boolean>(false)
    const [bubbleEnd, setBubbleEnd] = React.useState<React.Node>(null)

    React.useLayoutEffect(() => {
        const notifyViewEvent: NotifyViewEvent = createEvent();
        const stopEvent = createEvent<void>()

        const unsubscribe: Array<() => void> = []
        unsubscribe.push(notifyViewEvent.watch(viewData => {
            setData(viewData);
        }))
        unsubscribe.push(stopEvent.watch(() => {
            setStopped(true)
        }))

        runChat(page, {
            notifyView: notifyViewEvent,
            stopEvent,
            pending,
        });

        return () => {
            unsubscribe.forEach(e => e())
        }
    }, [setData, setStopped, runChat, page, pending]);

    if (!data) {
        return null
    }

    const {dialog, input} = data
    
    return <>
        <Dialog
            views={dialog}
            pending={pending?.pending}
            bubbleEnd={bubbleEnd}
            bubbleContainer={bubbleContainer}
        />
        <context.Provider value={setBubbleEnd}>
            {!stopped && input && <input.component {...input.props} key={dialog.size} />}
        </context.Provider>
    </>
}