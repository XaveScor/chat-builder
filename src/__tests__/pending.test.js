/* @flow */
import * as React from 'react'
import { createPage, createPending, input, Stop, questionBubble, useChatBuilder, createInput } from '..'
import { Chat } from '../ReactChat'
import * as renderer from 'react-test-renderer'
import { delay } from '../common'
import { createEvent } from '../event'
import { createChatMock, Mutex } from './common'

const pending = createPending({
	component: () => 'pending',
})

const pendingInput = createInput({
	component: () => 'pending input',
})

const pendingConfig = {
	pending,
	pendingTimeout: 50,
	input: pendingInput,
}

it('ConformsForm', async () => {
	const startPage = createPage(async () => {
		await delay(100)
		return {
			steps: [
				{
					question: questionBubble,
					questionProps: {
						question: 'question',
					},
					input,
				},
			],
			nextPage: Stop,
		}
	})

	const start = createEvent<void>()
	const mutex = new Mutex()
	function rerender(count) {
		mutex.release()
	}
	const chatMock = createChatMock(rerender, start)
	const tree = renderer.create(<Chat runChat={chatMock} page={startPage} pending={pendingConfig} />)

	renderer.act(() => {
		start()
	})

	const resOnlyInputChange = tree.toJSON()
	expect(resOnlyInputChange).toMatchSnapshot()

	await mutex.wait()

	const resShowPending = tree.toJSON()
	expect(resShowPending).toMatchSnapshot()

	await mutex.wait()

	const resHidePending = tree.toJSON()
	expect(resHidePending).toMatchSnapshot()
})

it('useChatBuilder', async () => {
	const startPage = createPage<void>(async () => {
		await delay(100)
		return {
			steps: [
				{
					question: questionBubble,
					questionProps: {
						question: 'question',
					},
					input,
				},
			],
			nextPage: Stop,
		}
	})

	const Form = ({ chatMock }) => {
		const Chat = useChatBuilder(startPage, {
			pending: pendingConfig,
		})

		return <Chat runChat={chatMock} />
	}

	const start = createEvent<void>()
	const mutex = new Mutex()
	function rerender(count) {
		mutex.release()
	}
	const chatMock = createChatMock(rerender, start)

	const tree = renderer.create(<Form chatMock={chatMock} />)
	renderer.act(() => {
		start()
	})

	const resOnlyInputChange = tree.toJSON()
	expect(resOnlyInputChange).toMatchSnapshot()

	await mutex.wait()

	const resShowPending = tree.toJSON()
	expect(resShowPending).toMatchSnapshot()

	await mutex.wait()

	const resHidePending = tree.toJSON()
	expect(resHidePending).toMatchSnapshot()
})
