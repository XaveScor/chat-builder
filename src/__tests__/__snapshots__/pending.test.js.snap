// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConformsForm 1`] = `"pending input"`;

exports[`ConformsForm 2`] = `
Array [
  "pending",
  "pending input",
]
`;

exports[`ConformsForm 3`] = `
Array [
  <p
    className="robot"
  >
    question
  </p>,
  <input
    onChange={[Function]}
  />,
  <input
    type="submit"
  />,
]
`;

exports[`useChatBuilder 1`] = `"pending input"`;

exports[`useChatBuilder 2`] = `
Array [
  "pending",
  "pending input",
]
`;

exports[`useChatBuilder 3`] = `
Array [
  <p
    className="robot"
  >
    question
  </p>,
  <input
    onChange={[Function]}
  />,
  <input
    type="submit"
  />,
]
`;
